docs/compilation/performance_tips.md

```diff
@@ -0,0 +1,505 @@
+    b = np.array([[1, x], [z, y]])
+    return fhe.array(a @ b)
+
+inputset = fhe.inputset(fhe.uint3, fhe.uint3, fhe.uint3)
+
+for dataflow_parallelize in [False, True]:
+    compiler = fhe.Compiler(f, {"x": "encrypted", "y": "encrypted", "z": "encrypted"})
+    circuit = compiler.compile(inputset, dataflow_parallelize=dataflow_parallelize)
+
+    start = time.time()
+    result = circuit.encrypt_run_decrypt(2, 1, 5)
+    end = time.time()
+    assert np.array_equal(result, f(2, 1, 5))
+
+    if dataflow_parallelize:
+        print(f"   with dataflow parallelize -> {(end - start):.03}s")
+    else:
+        print(f"without dataflow parallelize -> {(end - start):.03}s")
+```
+
+prints:
+
+```
+without dataflow parallelize -> 2.311s
+   with dataflow parallelize -> 1.893s
+```
+
+and the reason for that is:
+
+```
+// This is the generated MLIR for the operation
+// Without dataflow, every single line would be executed one after the other
+module {
+  func.func @main(%arg0: !FHE.eint<7>, %arg1: !FHE.eint<7>, %arg2: !FHE.eint<7>) -> tensor<2x2x!FHE.eint<7>> {
+    %c1_i2 = arith.constant 1 : i2
+    
+    // But if you look closely, you can see that this multiplication
+    %0 = "FHE.mul_eint_int"(%arg0, %c1_i2) : (!FHE.eint<7>, i2) -> !FHE.eint<7>
+    
+    // Is completely independent of this one and dataflow makes them run in parallel
+    %1 = "FHE.mul_eint"(%arg1, %arg2) : (!FHE.eint<7>, !FHE.eint<7>) -> !FHE.eint<7>
+    
+    // However this addition needs the first two operations, so it waits until both are done
+    %2 = "FHE.add_eint"(%0, %1) : (!FHE.eint<7>, !FHE.eint<7>) -> !FHE.eint<7>
+    
+    // Lastly, this multiplication is completely independent of the first three operations
+    // So its execution starts in parallel when execution starts
+    %3 = "FHE.mul_eint"(%arg0, %arg0) : (!FHE.eint<7>, !FHE.eint<7>) -> !FHE.eint<7>
+    
+    // Similar logic can be applied to the remaining operations
+    %4 = "FHE.mul_eint"(%arg1, %arg1) : (!FHE.eint<7>, !FHE.eint<7>) -> !FHE.eint<7>
+    %5 = "FHE.add_eint"(%3, %4) : (!FHE.eint<7>, !FHE.eint<7>) -> !FHE.eint<7>
+    %6 = "FHE.mul_eint_int"(%arg2, %c1_i2) : (!FHE.eint<7>, i2) -> !FHE.eint<7>
+    %c2_i3 = arith.constant 2 : i3
+    %7 = "FHE.mul_eint_int"(%arg2, %c2_i3) : (!FHE.eint<7>, i3) -> !FHE.eint<7>
+    %8 = "FHE.add_eint"(%6, %7) : (!FHE.eint<7>, !FHE.eint<7>) -> !FHE.eint<7>
+    %9 = "FHE.mul_eint"(%arg2, %arg0) : (!FHE.eint<7>, !FHE.eint<7>) -> !FHE.eint<7>
+    %10 = "FHE.mul_eint_int"(%arg1, %c2_i3) : (!FHE.eint<7>, i3) -> !FHE.eint<7>
+    %11 = "FHE.add_eint"(%9, %10) : (!FHE.eint<7>, !FHE.eint<7>) -> !FHE.eint<7>
+    %from_elements = tensor.from_elements %2, %5, %8, %11 : tensor<2x2x!FHE.eint<7>>
+    return %from_elements : tensor<2x2x!FHE.eint<7>>
+  }
+}
+```
+
+So, to summarize, dataflow analyzes the circuit to determine which parts of the circuit can be run at the same time, and tries to run as many operations as possible in parallel.
+
+{% hint style="info" %}
```

Comment:
warnings here more than info. Actually in an ideal world this should be decided by Concrete (runtime I guess) but that hard for now and maybe not the priority . Also there are an overhead of dataflow parallelism and if tasks are two small it can lead to slowdown too (wdyt @antoniupop ?).

Reply:
Yes, I've updated it to be a warning! And yes I'd love to get the opinion of @antoniupop and @andidr!